### ISSUE TITLE
Unable to apply Vector<Int> to Vector<T> on Java platform

### ISSUE THREAD

**Issue by Atry** [HaxeFoundation/haxe#3398]
``` haxe
import haxe.ds.Vector;

class VectorInt
{
    static function toVectorT<T>(v:Vector<T>) return v;

    public static function main()
    {
        var v:Vector<Int> = new Vector<Int>(0);
        toVectorT(v);
    }
}
```

```
$ haxe -java java -main VectorInt
haxelib run hxjava hxjava_build.txt --haxe-version 3200
D:\Program Files (x86)\Java\jdk1.7.0_55/bin/javac.exe "-sourcepath" "src" "-d" "obj" "-g:none" "@cmd"
src\haxe\root\VectorInt.java:36: 错误: 不可转换的类型
                haxe.root.VectorInt.toVectorT(((java.lang.Object[]) (((java.lang.Object[]) (v) )) ));
                                                                                           ^
  需要: Object[]
  找到:    int[]
1 个错误
Compilation error
Native compilation failed
Error: Build failed
```

**Comment by Simn** [comment]
I think this has to be a compiler error on Java. The types are simply not compatible at run-time, and using the wrapper element type would defeat the purpose of Vector (which is being fast).

**Comment by ncannasse** [comment]
We have similar error at runtime in HL.
The correct fix is to either inline or @:generic the function.

**Comment by Simn** [comment]
We should have some `@:invariant` metadata on these type parameters, but I still don't know how to implement that properly.

**Comment by ncannasse** [comment]
Yes, not sure how to get to the compile time error there. Most likely some late check like null for basic types

**Comment by Simn** [comment]
What we could do is have something like a default constraint on the type parameter, so that when you have an unconstrained `Vector<T>` it is actually `Vector<T:{}>`. I'm not sure how to handle that in syntax though, and the error message would probably be very confusing.

**Comment by ncannasse** [comment]
That's not very good semantic-wise either. You want T to be anything, not
any object on static platforms. And it works well with either inline or
generic.

**Comment by Simn** [comment]
The Java target will be put in "regression fixes only" state with the Haxe 4.3.0 release. See https://github.com/HaxeFoundation/haxe/wiki/Target-Support-Level#java-and-c-since-430 for additional information. We apologize for any inconvenience this causes!
